lib/iex/lib/iex/pry.ex

```diff
@@ -8,37 +19,64 @@ defmodule IEx.Pry do
   `IEx.pry/1` as a macro. This function expects the binding (from
   `Kernel.binding/0`) and the environment (from `__ENV__/0`).
   """
-  def pry(binding, env) do
+  def pry(binding, %Macro.Env{file: file, line: line} = env) do
+    self = self()
     opts = [binding: binding, dot_iex_path: "", env: env, prefix: "pry"]
-    meta = "#{inspect self()} at #{Path.relative_to_cwd(env.file)}:#{env.line}"
-    desc =
-      case whereami(env.file, env.line, 2) do
+    location = "#{Path.relative_to_cwd(file)}:#{line}"
+
+    whereami =
+      case whereami(file, line, 2) do
         {:ok, lines} -> [?\n, ?\n, lines]
-        :error -> ""
+        :error -> []
       end
 
-    res = IEx.Server.take_over("Request to pry #{meta}#{desc}", opts)
+    # If we are the current evaluator, it is because we just
+    # reached a pry/breakpoint and the user hit continue().
+    # In both cases, we are safe to print and the request will
+    # suceed.
+    request =
+      case IEx.Server.evaluator do
+        {^self, _} ->
+          IO.puts IEx.color :eval_interrupt, "Break reached: #{location}#{whereami}"
+          "Prying #{inspect self} at #{location}"
+        _ ->
+          "Request to pry #{inspect self} at #{location}#{whereami}"
+      end
 
-    # We cannot use colors because IEx may be off.
-    case res do
+    # We cannot use colors because IEx may be off
+    case IEx.Server.take_over(request, opts) do
+      :ok ->
+        :ok
       {:error, :no_iex} ->
         extra =
-          case :os.type do
-            {:win32, _} -> " If you are Windows, you may need to start IEx with the --werl flag."
-            _           -> ""
+          if match?({:win32, _}, :os.type) do
+            " If you are using Windows, you may need to start IEx with the --werl flag."
+          else
+            ""
           end
-        IO.puts :stdio, "Cannot pry #{meta}. Is an IEx shell running?" <> extra
-      _ ->
-        :ok
+        IO.puts :stdio, "Cannot pry #{inspect self} at #{location}. Is an IEx shell running?" <> extra
+        {:error, :no_iex}
+      {:error, _} = error ->
+        error
     end
+  end
 
-    res
+  def pry(binding, opts) when is_list(opts) do
+    vars = for {k, _} when is_atom(k) <- binding, do: {k, nil}
+    pry(binding, %{:elixir.env_for_eval(opts) | vars: vars})
   end
 
   @doc """
-  Formats the location for whereami prying.
+  Formats the location for `whereami/3` prying.
+
+  It receives the `file`, `line` and the snippet `radius` and
+  returns `{:ok, lines}`, where lines is a list of chardata
+  containing each formatted line, or `:error`.
+
+  The actual line is especially formatted in bold.
   """
-  def whereami(file, line, radius) do
+  def whereami(file, line, radius)
+      when is_binary(file) and is_integer(line) and is_integer(radius) and radius > 0 do
```

Comment:
Great idea. There is no hashing though. But something we could do for now is to print `mod.fun/arity` we are debugging. That will help with making it clearer in many cases.